iroh/examples/custom-protocol.rs

```diff
@@ -55,66 +128,157 @@ async fn main() -> Result<()> {
-            .connect_by_node_id(&remote_node_id, EXAMPLE_ALPN)
-            .await?;
-        let mut recv_stream = conn.accept_uni().await?;
-        let hash_bytes = recv_stream.read_to_end(32).await?;
-        let hash = iroh::blobs::Hash::from_bytes(hash_bytes.try_into().unwrap());
-        println!("received hash: {hash}");
-        self.client
-            .blobs()
-            .download(hash, remote_node_id.into())
-            .await?
-            .await?;
-        println!("blob downloaded");
-        let content = self.client.blobs().read_to_bytes(hash).await?;
-        let message = String::from_utf8(content.to_vec())?;
-        println!("blob content: {message}");
-        Ok(())
+    /// Query a remote node, download all matching blobs and print the results.
+    pub async fn query_remote(&self, node_id: NodeId, query: &str) -> Result<Vec<Hash>> {
+        // Establish a connection to our node.
+        // We use the default node discovery in iroh, so we can connect by node id without
+        // providing further information.
+        let conn = self.endpoint.connect_by_node_id(&node_id, ALPN).await?;
+
+        // Open a bi-directional in our connection.
+        let (mut send, mut recv) = conn.open_bi().await?;
+
+        // Send our query.
+        send.write_all(query.as_bytes()).await?;
+
+        // Finish the send stream, signalling that no further data will be sent.
+        // This makes the `read_to_end` call on the accepting side terminate.
+        send.finish().await?;
+
+        // In this example, we simply collect all results into a vector.
+        // For real protocols, you'd usually want to return a stream of results instead.
+        let mut out = vec![];
+
+        // The response is sent as a list of 32-byte long hashes.
+        // We simply read one after the other into a byte buffer.
+        let mut hash_bytes = [0u8; 32];
+        loop {
+            // Read 32 bytes from the stream.
+            match recv.read_exact(&mut hash_bytes).await {
+                // FinishedEarly means that the remote side did not send further data,
+                // so in this case we break our loop.
+                Err(quinn::ReadExactError::FinishedEarly) => break,
+                // Other errors are connection errors, so we bail.
+                Err(err) => return Err(err.into()),
+                Ok(_) => {}
+            };
+            // Upcast the raw bytes to the `Hash` type.
+            let hash = Hash::from_bytes(hash_bytes);
+            // Download the content via iroh-blobs.
+            self.blobs.download(hash, node_id.into()).await?.await?;
+            // Add the blob to our local database.
+            self.add_to_index(hash).await?;
+            out.push(hash);
+        }
+        Ok(out)
+    }
+
+    /// Query the local database.
+    ///
+    /// Returns the list of hashes of blobs which contain `query` literally.
+    pub async fn query_local(&self, query: &str) -> Result<Vec<Hash>> {
```

Comment:
```suggestion
    pub fn query_local(&self, query: &str) -> Vec<Hash> {
```

Needed to do this to fix some type error.

Reply:
uh yea thanks, this was a refactor leftover